drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_decoder.c

```diff
@@ -0,0 +1,551 @@
+	[LSM6DSV16X_XL_BATCHED_AT_480Hz] = UINT32_C(1000000000) / 480,
+	[LSM6DSV16X_XL_BATCHED_AT_960Hz] = UINT32_C(1000000000) / 960,
+	[LSM6DSV16X_XL_BATCHED_AT_1920Hz] = UINT32_C(1000000000) / 1920,
+	[LSM6DSV16X_XL_BATCHED_AT_3840Hz] = UINT32_C(1000000000) / 3840,
+	[LSM6DSV16X_XL_BATCHED_AT_7680Hz] = UINT32_C(1000000000) / 7680,
+};
+
+static uint32_t gyro_period_ns[] = {
+	[LSM6DSV16X_GY_BATCHED_AT_1Hz875] = UINT32_C(1000000000000) / 1875,
+	[LSM6DSV16X_GY_BATCHED_AT_7Hz5] = UINT32_C(1000000000000) / 7500,
+	[LSM6DSV16X_GY_BATCHED_AT_15Hz] = UINT32_C(1000000000) / 15,
+	[LSM6DSV16X_GY_BATCHED_AT_30Hz] = UINT32_C(1000000000) / 30,
+	[LSM6DSV16X_GY_BATCHED_AT_60Hz] = UINT32_C(1000000000) / 60,
+	[LSM6DSV16X_GY_BATCHED_AT_120Hz] = UINT32_C(1000000000) / 120,
+	[LSM6DSV16X_GY_BATCHED_AT_240Hz] = UINT32_C(1000000000) / 240,
+	[LSM6DSV16X_GY_BATCHED_AT_480Hz] = UINT32_C(1000000000) / 480,
+	[LSM6DSV16X_GY_BATCHED_AT_960Hz] = UINT32_C(1000000000) / 960,
+	[LSM6DSV16X_GY_BATCHED_AT_1920Hz] = UINT32_C(1000000000) / 1920,
+	[LSM6DSV16X_GY_BATCHED_AT_3840Hz] = UINT32_C(1000000000) / 3840,
+	[LSM6DSV16X_GY_BATCHED_AT_7680Hz] = UINT32_C(1000000000) / 7680,
+};
+
+#if defined(CONFIG_LSM6DSV16X_ENABLE_TEMP)
+static uint32_t temp_period_ns[] = {
+	[LSM6DSV16X_TEMP_BATCHED_AT_1Hz875] = UINT32_C(1000000000000) / 1875,
+	[LSM6DSV16X_TEMP_BATCHED_AT_15Hz] = UINT32_C(1000000000) / 15,
+	[LSM6DSV16X_TEMP_BATCHED_AT_60Hz] = UINT32_C(1000000000) / 60,
+};
+#endif
+
+static int lsm6dsv16x_get_shift(enum sensor_channel channel,
+				int accel_fs, int gyro_fs, int8_t *shift)
+{
+	int ret = 0;
+
+	switch (channel) {
+	case SENSOR_CHAN_ACCEL_XYZ:
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+		switch (accel_fs) {
+		case LSM6DSV16X_DT_FS_2G:
+			*shift = 5;
+			break;
+		case LSM6DSV16X_DT_FS_4G:
+			*shift = 6;
+			break;
+		case LSM6DSV16X_DT_FS_8G:
+			*shift = 7;
+			break;
+		case LSM6DSV16X_DT_FS_16G:
+			*shift = 8;
+			break;
+		default:
+			LOG_ERR("bad accel fs = %d", accel_fs);
+			return -EINVAL;
+		}
+		break;
+
+	case SENSOR_CHAN_GYRO_XYZ:
+	case SENSOR_CHAN_GYRO_X:
+	case SENSOR_CHAN_GYRO_Y:
+	case SENSOR_CHAN_GYRO_Z:
+		switch (gyro_fs) {
+		case LSM6DSV16X_DT_FS_125DPS:
+			*shift = 2;
+			break;
+		case LSM6DSV16X_DT_FS_250DPS:
+			*shift = 3;
+			break;
+		case LSM6DSV16X_DT_FS_500DPS:
+			*shift = 4;
+			break;
+		case LSM6DSV16X_DT_FS_1000DPS:
+			*shift = 5;
+			break;
+		case LSM6DSV16X_DT_FS_2000DPS:
+			*shift = 6;
+			break;
+		case LSM6DSV16X_DT_FS_4000DPS:
+			*shift = 7;
+			break;
+		default:
+			LOG_ERR("bad gyro fs = %d", gyro_fs);
+			return -EINVAL;
+		}
+		break;
+
+#if defined(CONFIG_LSM6DSV16X_ENABLE_TEMP)
+	case SENSOR_CHAN_DIE_TEMP:
+		*shift = 16; /* specification Q15.16 */
+		break;
+#endif
+	default:
+		LOG_ERR("bad channel num = %d", channel);
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int lsm6dsv16x_convert_raw_to_q31(const struct lsm6dsv16x_decoder_header *hdr,
```

Comment:
Converting from the sensors sample representation, almost always a fixed point value with its own scale, to the larger q31 format should be a simple multiplier provided by a lookup table.

E.g.

``` c
static inline void lsm6dsv16x_accel_to_q31(int16_t accel_raw, int accel_fs, q31_t *out)
{
    *out = accel_raw * ACCEL_CONVERSION[accel_fs];
}
```

See issue #76968

Reply:
ok, done. Please consider it as just a proposal that should address #76968